bin/changelog.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python
+import os
+import sys
+import re
+import requests
+import subprocess
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+from collections import defaultdict
+
+HEADERS = {
+    'M': 'Major changes',
+    'b': 'Backwards compatibility breaks and deprecations',
+    'n': 'New features',
+    'm': 'Minor changes'
+}
+PREFIX = '* '
+SUFFIX = ' ([#{pr_number}](../pull/{pr_number}))\n'
+
+def red(text):
+    return "\033[31m%s\033[0m" % text
+
+def green(text):
+    return "\033[32m%s\033[0m" % text
+
+def blue(text):
+    return "\033[34m%s\033[0m" % text
+
+def get_build_information():
+    """
+    Check event type and return PR number
+    """
+    event_type = os.environ['TRAVIS_EVENT_TYPE']
+    if event_type == 'push':
+        match = re.match(r'Merge pull request #(\d+)',
+            os.environ['TRAVIS_COMMIT_MESSAGE']);
+        if not match:
+            sys.exit(red('Cannot find pull request number!'))
+        return match.group(1)
+    elif event_type == 'pull_request':
+        return os.environ['TRAVIS_PULL_REQUEST']
+    elif event_type == 'cron' or event_type == 'api':
+        print(green('Not a push or PR build, skipping changelog'))
+        sys.exit()
+    else:
+        sys.exit(red('Unknown event type!'))
+
+def request_https_get(url):
+    """
+    Make HTTPS GET request and return response
+    """
+    s = requests.Session()
+    retry = Retry(total=5, read=5, connect=5, backoff_factor=0.1,
+        status_forcelist=(500, 502, 503, 504))
+    adapter = HTTPAdapter(max_retries=retry)
+    s.mount('https://', adapter)
+    r = s.get(url)
+    r.raise_for_status()
+    return r
+
+def get_pr_desc(pr_number):
+    """
+    Retrieve pull request description using GitHub API
+    """
+    r = request_https_get('https://api.github.com/repos/' +
+        os.environ['TRAVIS_REPO_SLUG'] + '/pulls/' + pr_number)
+    return r.json()['body']
+
+def get_changelog(data):
+    """
+    Parse changelogs from a string
+    """
+    changelogs = defaultdict(list)
+    start = False
+    for line in data.splitlines():
+        if 'This PR changes' in line:
```

Comment:
I would prefer that it start with `[CHANGELOG START]`. I found that the target text was not always a natural way for me to begin the discussion of the PR. Also, doesn't what you have here treat that opening line like a message to be parsed since it ends with a ":"? Perhaps it would be better to `continue` after detecting the start.

Reply:
@smichr `This PR changes` is a newly added level 2 heading on the pull request template.